### ISSUE TITLE
Python: implement typemap for the Python backend

### ISSUE THREAD

**Issue by tasn** [Dushistov/flapigen-rs#340]
As suggested in https://github.com/Dushistov/flapigen-rs/pull/236#issuecomment-665161413 unfortunately type maps aren't supported in Python.
The original PR indicated that it may be possible using the rust_cpython traits though unfortunately that's not possible because you can't define traits for types that weren't created in your crate.

The type mapping in the Java backend work great, maybe these could be implemented for the python backend too? I tried doing it myself but I'm a bit lost to be honest. I'm still new to the macro-magic that's involved in making this lib and have never used syn.

Thanks!

**Comment by AdamRzepka** [comment]
I'd need to take a look at the Python backend code again to get the idea how much work this require. The good news is, I've implemented C# backend (I'll fill the PR soon) in the fashion of type maps, so I better understand now, how they work. The bad news - I definitely won't have a time in a nearest month to do any major improvements.

Can you describe your use case more precisely? Maybe there's  a workaround for a Python backend.

**Comment by tasn** [comment]
I managed to work around almost all of my issues by just implementing wrappers for my classes, though that's not ideal, obviously. :)
My main issues are:
1. Returning bytes - the API currently returns a list of numbers as a byte. It's almost the same in Python so most people won't notice it, but still annoying.
2. Accepting Option<&T> and Vec<&T> - we currently had to implement a lot of code manually and have a lot of copies (#339)
3. Exceptions - currently the exceptions are just one type with different strings, would be great to be able to map them to real Python exception types so we have different classes for exceptions.
4. `@property` - would be great if we could denote a function as a setter and a function as a getter. Though I just thought about it, maybe we can use foreign code to add this? Not sure.

Thanks a lot!

**Comment by tasn** [comment]
Having actually used the bindings now, one of the more annoying things about them is the lack of `@property` support. It really makes the API ugly. I think I'll end up just wrapping the objects to have a nicer API on top.

**Comment by tasn** [comment]
I ended up wrapping all of my API manually in order to fix the lack of `@property` support, the `bytes` return type for `Vec<u8>` and a few other things.
Please note that I had to resort to having `_inner` objects because the generated Python classes can't be inherited from for some reason (libcypthon limitation? no idea).
The main thing I haven't worked around yet, just because it'll be really ugly and manual is the exception handling. I really don't want to wrap each function with a catch that will try to string match the error type and raise its own error. :|
Here's the code: https://github.com/etesync/etebase-py/blob/master/etebase/__init__.py
The glue code is also in the same repo.

**Comment by tasn** [comment]
@AdamRzepka, did you get a chance to look at it? Are my explanations clear?

Thanks again!

**Comment by AdamRzepka** [comment]
Your explanations are clear. Unfortunately, I still haven't a chance to
look into this. Sorry about that.

wt., 8 wrz 2020 o 14:40 Tom Hacohen <notifications@github.com> napisał(a):

**Comment by tasn** [comment]
No worries, was just making sure there was nothing on my end that was missing. Thanks again!

**Comment by tasn** [comment]
Just following up on this.

I realise you may be too busy to take a look, but do you still have an idea on how I can at least handle throwing different Exception types?

Thanks!

**Comment by AdamRzepka** [comment]
You'll need to alter `macroslib/src/python/mod.rs` in this library. However, I believe, this is should be quite simple change. AFAIR, there'll be following steps:
- In function `generate_conversion_for_return`, there's a block starting with `} else if let Some((inner_ok, _inner_err)) = ast::if_result_return_ok_err_types(&rust_type) {`. It is responsible for `Result` handling. Change the code, so it uses custom `inner_err` type, instead of generic one.
- Add single `py_exception!` macro call for each type, that is used there as an error.

**Comment by tasn** [comment]
Thanks!
Correct me if I'm wrong, but I think the best way to handle it would be to define the error types in Python and then here just either call a function or the constructor of that type that translates the Rust Error type I use. What do you think @Dushistov? I would like to make the code generic so that it can be merged into flapigen for everyone's benefit (though if not possible, happy to just hack it here for my own usage).
If so how do I go about doing this?

Thanks again!

**Comment by AdamRzepka** [comment]
Yeah, these error types should be defined in Python. However, I think, that this should be done automatically, through the `foreign_class` macro (just like any other exported type).

**Comment by tasn** [comment]
I can see how I can create new error types from the module, by going to:
```
            mod py_error {
                py_exception!(#module_name, Error);
            }

            py_module_initializer!(#module_name, #module_init, #module_py_init, |py, m| {
                m.add(py, "Error", py_error::Error::type_object(py))?;
                #(#module_initialization_code)*
                Ok(())
            });
```
Though that's probably not what I want. What I probably want is be able to define the Python exceptions in Python and then call them from:
```
                    Err(err_inner) => return Err(cpython::PyErr::new::<super::py_error::Error, _>(
                        py,
                        swig_collect_error_message(&err_inner)
                    )),
```
No?

But I can't figure out how to do that, as it accepts a Rust type here, and I can't figure out how to access python code.

**Comment by tasn** [comment]
Hey, just following up on this.

Btw, any eta for the C# bindings? I'm looking forward to giving that a spin. :)
